crates/consensus/src/transaction/eip4844.rs

```diff
@@ -3,13 +3,303 @@ use alloy_eips::{eip2930::AccessList, eip4844::DATA_GAS_PER_BLOB};
 use alloy_network::{Signed, Transaction};
 use alloy_primitives::{keccak256, Bytes, ChainId, Signature, B256, U256};
 use alloy_rlp::{length_of_length, BufMut, Decodable, Encodable, Header};
-use std::mem;
+use c_kzg::{Blob, Bytes48, KzgCommitment, KzgProof, KzgSettings};
+use sha2::Digest;
+use std::{mem, ops::Deref};
+
+/// An error that can occur when validating a [TxEip4844].
+#[derive(Debug, thiserror::Error)]
+pub enum BlobTransactionValidationError {
+    /// Proof validation failed.
+    #[error("invalid KZG proof")]
+    InvalidProof,
+    /// An error returned by [`c_kzg`].
+    #[error("KZG error: {0:?}")]
+    KZGError(#[from] c_kzg::Error),
+    /// The inner transaction is not a blob transaction.
+    #[error("unable to verify proof for non blob transaction: {0}")]
+    NotBlobTransaction(u8),
+    /// The versioned hash is incorrect.
+    #[error("wrong versioned hash: have {have}, expected {expected}")]
+    WrongVersionedHash {
+        /// The versioned hash we got
+        have: B256,
+        /// The versioned hash we expected
+        expected: B256,
+    },
+}
 
 /// [EIP-4844 Blob Transaction](https://eips.ethereum.org/EIPS/eip-4844#blob-transaction)
 ///
-/// A transaction with blob hashes and max blob fee
+/// A transaction with blob hashes and max blob fee.
+/// It can either be a standalone transaction, mainly seen when retrieving historical transactions,
+/// or a transaction with a sidecar, which is used when submitting a transaction to the network and
+/// when receiving and sending transactions during the gossip stage.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TxEip4844 {
+    /// A standalone transaction with blob hashes and max blob fee.
+    TxEip4844(BlobTx),
+    /// A transaction with a sidecar, which contains the blob data, commitments, and proofs.
+    TxEip4844WithSidecar((BlobTx, BlobTransactionSidecar)),
```

Comment:
I'd prefer a standalone type that has the tx + blob, the tuple here makes it harder to work with